roles/tests-integration/tests/pool_integration.rs

```diff
@@ -0,0 +1,66 @@
+use std::{net::SocketAddr, str::FromStr, time::Duration};
+
+use tokio::time::sleep;
+
+use crate::common::{get_available_port, is_port_open};
+
+mod common;
+
+#[tokio::test]
+async fn success_pool_template_provider_connection() {
+    let template_provider_port = get_available_port().unwrap();
+    let template_provider = common::TemplateProvider::start(template_provider_port);
+    template_provider.generate_blocks(16);
+    let pool_port = get_available_port().unwrap();
```

Comment:
this is causing some confusion for me

it seems that `get_available_port` binds to a random port, which at a first glance seems like a clever approach

but then, we give this string to some constructors, and eventually, we are going to have `Pool::accept_incoming_connection`, which will bind again to the same socket:

https://github.com/stratum-mining/stratum/blob/708e5e4ecb1f67d4e629b68dd4965271ced389c4/roles/pool/src/lib/mining_pool/mod.rs#L415-L420

Reply:
hmm maybe we're safe due to the fact that we only call `accept` from inside `accept_incoming_connection`, but never before

anyways I'll leave this here for a sanity check

Reply:
quick test:
```rust
fn main() {
    let a = std::net::TcpListener::bind("127.0.0.1:8080");
    println!("{:?}", a);

    let b = std::net::TcpListener::bind("127.0.0.1:8080");
    println!("{:?}", b);
}
```

this prints:
```
Ok(TcpListener { addr: 127.0.0.1:8080, fd: 3 })
Err(Os { code: 98, kind: AddrInUse, message: "Address already in use" })
```

and:
```
fn main() {
    let a = std::net::TcpListener::bind("127.0.0.1:8080");
    println!("{:?}", a);

    drop(a);

    let b = std::net::TcpListener::bind("127.0.0.1:8080");
    println!("{:?}", b);
}
```

prints:
```
Ok(TcpListener { addr: 127.0.0.1:8080, fd: 3 })
Ok(TcpListener { addr: 127.0.0.1:8080, fd: 3 })
```

so we're only safe as long as `pool_port` is dropped before we call `pool.start()`

Reply:
do we REALLY need `get_available_port`? why not just leave it to the user to choose unique ports?

if feels like the tradeoffs here could eventually turn into a headache down the line

Reply:
I see your point. For some reason I thought cargo executed unit tests sequentially, one after the other.

But I just did some local experiments and indeed, it seems they are executed in parallel.

also, I just realized that `get_available_port` only returns the `u16` representing the port, but not the `TcpListener` object itself... so we only briefly bind to the port, drop it, and then later on bind again inside the actual role

now things are more clear in my head, thanks for the clarification!

but now I'm wondering: aren't we leaving some room for undeterministic behavior?

imagine that two separate tests are being executed in parallel, and by bad luck `get_available_port` assigns the same port to both, because of this window of time where we only have a `u16` but no actual socket binding to the port

Reply:
This makes sense @plebhash .
Let me point out my concerns.
When we use the `#[tokio::test]` macro inside a module, Tokio gives the whole module a single runtime. In these tests, we're running tasks for different roles, like servers, which might block the runtime. This can lead to tests sharing threads and system resources, which can slow things down, even though the tests are running in parallel.
A better way, like in the MG tests, would be to split the tests into different files, with each file focused on specific scenarios. This way, Tokio can assign a separate runtime to each test file, preventing shared resource issues. If we did this, it would also make port allocation easier, allowing us to reuse the same port across tests without any conflicts.
Do let me know what are your thoughts on this? cc: @plebhash @jbesraa

Reply:
> we're running tasks for different roles, like servers, which might block the runtime.

can you elaborate? how would it block the runtime?

Reply:
like Pool, JDC, JDS...blocking the runtime threads..

Reply:
hmm maybe I need to dive deeper into Tokio runtimes to get a better grasp of what you are saying

but I wouldn't oppose to having tests living in separate files, I can also see other benefits in that

the downside is slower execution, but I would always prioritize determinism over that

Reply:
> This makes sense @plebhash . Let me point out my concerns. When we use the `#[tokio::test]` macro inside a module, Tokio gives the whole module a single runtime. In these tests, we're running tasks for different roles, like servers, which might block the runtime. This can lead to tests sharing threads and system resources, which can slow things down, even though the tests are running in parallel. A better way, like in the MG tests, would be to split the tests into different files, with each file focused on specific scenarios. This way, Tokio can assign a separate runtime to each test file, preventing shared resource issues. If we did this, it would also make port allocation easier, allowing us to reuse the same port across tests without any conflicts. Do let me know what are your thoughts on this? cc: @plebhash @jbesraa

I dont get the shared resources issues TBH, and I am not sure if it is something we should care about or is it a tokio concern. In general if your concern is performance, then option b, separating each test into a separate file wouldn't really solve that.

Reply:
I asked @Shourya742 to elaborate on his concerns around shared resources:

> When running async tests or functions in Rust, a runtime is required because Rust doesn't provide a built-in async runtime. We use tokio for this. We use macros like #[tokio::main] or #[tokio::test] to initialize the Tokio runtime, which manages tasks and threads. Tokio creates a thread pool, and these threads are used to run asynchronous tasks (often referred to as green threads). Rust futures, which represent async operations, are lazy, meaning they won't execute until they're awaited. The Tokio runtime has a task queue, and an executor is responsible for polling the futures to drive them to completion. In our case, when we use #[tokio::test] to run multiple async tests, the tests might share the same runtime, and by extension, the same thread pool. This means that the resources available for executing tasks are shared across all tests.

> Each of our tests spawns multiple roles, which in turn create many async tasks. Although these tasks appear to be asynchronous, the sheer number of them can overload the runtime's task queue. This causes delays because the executor has to manage and poll a large number of futures. Even if a task like an async sleep is simple, if there are too many other tasks in the queue, it could end up waiting longer than expected before being polled.

> For instance, suppose we have six tasks in total: five async tasks that perform quick, non-blocking operations and one task with an async sleep of 10 seconds. If our thread pool only has a single thread, all these tasks will be queued. While the sleep task is set for 10 seconds, it might not start exactly after 10 seconds because the other tasks are also competing for the same thread. The runtime's task queue can delay the polling of the sleep task, making it wait longer than the expected 10 seconds. This kind of delay grows more pronounced when more tasks are spawned by each role, ultimately slowing down the entire test process.


> In our scenario, running multiple tests in the same file/module results in many tasks being spawned, which bloats the runtime's task queue. This slows down the overall execution, especially since our roles spawn many tasks, making the testing process less efficient.

Reply:
Yup sure!!!

Reply:
Let me review the PR.. Then we are good to go..